controllers/vmware/vspherecluster_reconciler.go

```diff
@@ -369,9 +370,68 @@ func (r *ClusterReconciler) VSphereMachineToCluster(ctx context.Context, o clien
 	}}
 }
 
+// ZoneToVSphereClusters adds reconcile requests for VSphereClusters when Zone has an event.
+func (r *ClusterReconciler) ZoneToVSphereClusters(ctx context.Context, o client.Object) []reconcile.Request {
+	log := ctrl.LoggerFrom(ctx)
+
+	zone, ok := o.(*topologyv1.Zone)
+	if !ok {
+		log.Error(nil, fmt.Sprintf("Expected a Zone but got a %T", o))
+		return nil
```

Comment:
```suggestion
		panic(fmt.Sprintf("Expected a Zone but got a %T", o))
```

We should never hit this caes, except we did something wrong so we can directly panic instead.

Reply:
In all other cases in CAPV we just log this error. I'm fine with panic'ing here for now

Should we open an issue in CAPI & CAPV to just move all of this stuff to generic's and then there is no need for the cast?

Reply:
Referring CAPI: we panic there :D

Reply:
I'm finew ith both! I'll create an issue for going to generics.

Reply:
xref

Reply:
Yeah I saw that it's handled differently between CAPI & CAPV :)